pkg/sql/distsql_plan_backfill.go

```diff
@@ -83,3 +100,55 @@ func (dsp *DistSQLPlanner) createBackfillerPhysicalPlan(
 	dsp.FinalizePlan(planCtx, p)
 	return p, nil
 }
+
+// createIndexBackfillerMergePhysicalPlan generates a plan consisting
+// of index merger processors, one for each node that has spans that
+// we are reading. The plan is finalized.
+func (dsp *DistSQLPlanner) createIndexBackfillerMergePhysicalPlan(
+	planCtx *PlanningCtx, spec execinfrapb.IndexBackfillMergerSpec, spans [][]roachpb.Span,
+) (*PhysicalPlan, error) {
+
+	var indexSpans []roachpb.Span
+	var spanIdxs []interface{}
+
+	for i := range spans {
+		for j := range spans[i] {
+			indexSpans = append(indexSpans, spans[i][j])
+			spanIdxs = append(spanIdxs, i)
+		}
+	}
+	spanPartitions, idxByPartition, err := dsp.PartitionSpansWithUserData(planCtx, indexSpans, spanIdxs)
```

Comment:
this all feels gross. I really don't like `PartitionSpansWithUserData`. Probably with as much code, and fewer allocations than you've infected this code with, you could use a data structure to look up the index:

```go
type spanAndIndex struct {
    roachpb.Span
    idx int
}

var _ interval.Interface = (*spanAndIndex)(nil)

func (si *spanAndIndex) Range() inteval.Range { return si.AsRange() }
func (si *spanAndIndex) ID() uintptr { return uintptr(unsafe.Pointer(si)) }

func (dsp *DistSQLPlanner) createIndexBackfillerMergePhysicalPlan(
	planCtx *PlanningCtx, spec execinfrapb.IndexBackfillMergerSpec, spans [][]roachpb.Span,
) (*PhysicalPlan, error) {
        var n int
	for _, sp := range spans {
		for range sp {
		        n++
		}
	}
	indexSpans := make([]roachpb.Span, 0, n)
        spanIdxs := make([]spanAndIdxs, 0, n)
        spanIdxTree := interval.NewTree(interval.ExclusiveOverlapper)
        for i := range spans {
		for j := range spans[i] {
			indexSpans = append(indexSpans, spans[i][j])
			spanIdxs = append(spanIdxs, spanAndIndex{Span: spans[i][j], idx: i})
			spanIdxTree.Insert(&spanIdxs[len(spanIdxs)-1], true /* fast */)
		}
	}
	spanIdxTree.AdjustRanges()
	spanPartitions, err := dsp.PartitionSpansWithUserData(planCtx, indexSpans)
	// ...
	// now you can look up an index for a span like:
	getIndex := func(sp roachpb.Span) (idx int) {
	    if !spanIdxTree.DoMatching(func(i Interface) (done bool) {
	        i.(*spanAndIndex).idx
	        return false
	     }, sp.AsRange()) {
	        panic("didn't find it")
	     }
	     return idx
	}
	
```

Reply:
Thanks for this suggestion.  I've updated the PR to use this method. @rhu713 Mind taking a quick look at the last few commits to make sure you are OK with those changes.